src/libsyntax/parse/mod.rs

```diff
@@ -494,8 +494,17 @@ fn float_lit(s: &str, suffix: Option<Symbol>, diag: Option<(Span, &Handler)>)
                  -> Option<ast::LitKind> {
     debug!("float_lit: {:?}, {:?}", s, suffix);
     // FIXME #2252: bounds checking float literals is deferred until trans
-    let s = s.chars().filter(|&c| c != '_').collect::<String>();
-    filtered_float_lit(Symbol::intern(&s), suffix, diag)
+
+    // Strip underscores without allocating a new String unless necessary.
+    let s2;
+    let s = if s.chars().any(|c| c == '_') {
+        s2 = s.chars().filter(|&c| c != '_').collect::<String>();
```

Comment:
Hm, since underscores are in ASCII, couldn't we into_bytes, retain, and then String::from_utf8, maybe unchecked?

Reply:
That would work, but it's not clear to me that it would be any faster; it might even be slower because there are two allocations involved? (Because it converts to a vector, and then back to a new String?)

Besides, this is the cold path, so I'm satisfied with reusing the existing code, which until now had been considered good enough for the hot path :)

Reply:
Just FYI, [String::from_utf8](https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8) consumes the vector given to it, so no additional allocation would happen.